src/iteration_utils.jl

```diff
@@ -16,10 +16,37 @@ end
     end
 end
 
+# fast version for the case where J and sparsity have the same sparsity pattern
+@inline function _colorediteration!(Jsparsity::SparseMatrixCSC,vfx,colorvec,color_i,ncols)
+    @inbounds for col_index in 1:ncols
+        if colorvec[col_index] == color_i
+            @inbounds for spidx in nzrange(Jsparsity, col_index)
+                row_index = Jsparsity.rowval[spidx]
+                Jsparsity.nzval[spidx]=vfx[row_index]
+            end
+        end
+    end
+end
+
 #override default setting of using findstructralnz
 _use_findstructralnz(sparsity) = ArrayInterface.has_sparsestruct(sparsity)
 _use_findstructralnz(::SparseMatrixCSC) = false
 
+# test if J, sparsity are both SparseMatrixCSC and have the same size storage arrays,
+# if so, update J so they can share the same sparsity pattern
+_use_sparseCSC_common_sparsity!(J, sparsity) = false
+function _use_sparseCSC_common_sparsity!(J::SparseMatrixCSC, sparsity::SparseMatrixCSC)
+    common_sparsity = (length(J.colptr) == length(sparsity.colptr) &&
+            length(J.nzval) == length(sparsity.nzval))

```

Comment:
I'm not sure about this. This just means that they have the same data size. Now, this might be fine because I can't think of a reason for someone hitting this case on purpose: the reason for `sparsity` vs `J` is so that you can say keep `J` dense but still use color differentiation (for smaller matrices), or do partial/approximate coloring. But if the sparsity has the same sparsity as the matrix you're trying to fill, then partial/approximate coloring doesn't make sense... so I think this is fine. I'll merge it under this understanding but it's good to have it mentioned.

Reply:
Can there be a test on this behavior?

Reply:
Updated version simplifies this logic so that the fast path is only taken if `J` and `sparsity` have the same sparsity pattern (by comparing colptr and rowval).

There is an existing test (in coloring_tests.jl) for color differentiation writing to a dense `J`

This does reveal what I think might be pre-exisiting issues with accumulating components of a Jacobian into `J`, but I'm not sure what the use cases are here:
- in general, `J` is initialised to zero, and elements are set, not added (so this isn't useful to accumulate)
- if `J` is sparse, but with different stored elements to `sparsity` it's not clear to me what the correct behaviour should be:
  - error ? (instead of silently taking a slow path)
  - reset `J` to the pattern of `sparsity` ? (might make sense if the intent is to set `J`. Would allocate, but could be optimized)
  - keep `J` pattern, initialize to zero, and add new stored elements if needed ? (current behaviour)

Reply:
Yeah, the purpose I was thinking of was that, maybe you don't have just dense vs sparse at different parts of the code, but also do something like partial coloring so the sparsity may be in general Tridiagonal while the sparse matrix is not tridiagonal, i.e. using coloring on a subset so that you can use fewer `f` calculations and get an approximation to the full Jacobian. But if you do that, wouldn't you then want to just specialize the linear system solve on the approximate tridiagonal-ness? So I think this flexibility is somewhat left open as a research question. 

But if it's two sparse matrices, I would assume it would only make sense if one has a subset of zeros of the other.